Chime/Parser.m

```diff
@@ -0,0 +1,66 @@
+#import "Headers/Parser.h"
+#import "Headers/Utilities.h"
+#import <Foundation/Foundation.h>
+
+id parse(char *filepath) {
+  NSData *data = [[NSData alloc] initWithContentsOfFile:@(filepath)];
+
+  if (data == nil) {
+    NSLog(@"Could not read file \"%s\"", filepath);
+    return nil;
+  }
+
+  NSString *stringProgram =
+      [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+
+  NSArray *program = [stringProgram componentsSeparatedByString:@"\n"];
+  NSMapTable *labels = passOne(program);
+  return passTwo(labels, program);
+}
+
+BOOL isLineStartingWith(NSString *line, NSString *target) {
+  return [line hasPrefix:target];
+}
+
+BOOL isNumberLine(NSString *line) { return isLineStartingWith(line, @"#"); }
+
+BOOL isLabelLine(NSString *line) { return isLineStartingWith(line, @":"); }
+
+NSMapTable *passOne(NSArray *program) {
+  NSMapTable *labels = [[NSMapTable alloc] init];
+  NSEnumerator *programEnumerator = [program objectEnumerator];
+  id lineId;
+  NSUInteger counter = 0;
+  while (lineId = [programEnumerator nextObject]) {
+    if (isLabelLine(lineId)) {
+      NSArray *line = [lineId componentsSeparatedByString:@" "];
+      NSString *labelColon = [line objectAtIndex:0];
+      NSString *label = [labelColon substringFromIndex:1];
+      if (findInEnumerator([labels keyEnumerator], label)) {
```

Comment:
What does it mean to find the same label in the map again in the pass? Why is that a collision, I mean?

Reply:
Because finding it already there means that the programmer is using the same labels for two different things.